sky/provision/gcp/instance.py

```diff
@@ -72,6 +90,237 @@ def _wait_for_operations(
+    #  of instances dynamically. Then this case would not be an error.
+    if config.resume_stopped_nodes and len(exist_instances) > config.count:
+        raise RuntimeError('The number of running/stopped/stopping '
+                           f'instances combined ({len(exist_instances)}) in '
+                           f'cluster "{cluster_name}" is greater than the '
+                           f'number requested by the user ({config.count}). '
+                           'This is likely a resource leak. '
+                           'Use "sky down" to terminate the cluster.')
+
+    to_start_count = (config.count - len(running_instances) -
+                      len(pending_instances))
+
+    # Try to reuse previously stopped nodes with compatible configs
+    if config.resume_stopped_nodes and to_start_count > 0 and stopped_instances:
+        resumed_instance_ids = [n['name'] for n in stopped_instances]
+        if resumed_instance_ids:
+            for instance_id in resumed_instance_ids:
+                resource.start_instance(instance_id, project_id,
+                                        availability_zone)
+                resource.set_labels(project_id, availability_zone, instance_id,
+                                    labels)
+        to_start_count -= len(resumed_instance_ids)
+
+        if head_instance_id is None:
+            head_instance_id = resource.create_node_tag(
+                cluster_name,
+                project_id,
+                availability_zone,
+                resumed_instance_ids[0],
+                is_head=True,
+            )
+
+    if to_start_count > 0:
+        results = resource.create_instances(cluster_name, project_id,
+                                            availability_zone,
+                                            config.node_config, labels,
+                                            to_start_count,
+                                            head_instance_id is None)
+        if head_instance_id is None:
+            head_instance_id = results[0][1]
+        # FIXME: it seems that success could be False sometimes, but the instances
+        #   are started correctly.
+        created_instance_ids = [instance_id for success, instance_id in results]
+
+    return common.ProvisionRecord(provider_name='gcp',
+                                  region=region,
+                                  zone=availability_zone,
+                                  cluster_name=cluster_name,
+                                  head_instance_id=head_instance_id,
+                                  resumed_instance_ids=resumed_instance_ids,
+                                  created_instance_ids=created_instance_ids)
+
+
+def wait_instances(region: str, cluster_name: str,
+                   state: Optional[status_lib.ClusterStatus]) -> None:
+    """See sky/provision/__init__.py"""
+    # We already wait for the instances to be running in run_instances.
+    # So we don't need to wait here.
```

Comment:
Seems we only wait for the operator in `run_instances` but not the actual status. It seems causing the following error when running `sky launch --cloud gcp --cpus 2 -c test-new-gcp-2`
```
D 10-23 00:20:45 provisioner.py:485] ==================== System Setup After Provision ====================
D 10-23 00:20:45 provisioner.py:485] 
D 10-23 00:20:45 instance.py:55] handlers: [<class 'sky.provision.gcp.instance_utils.GCPComputeInstance'>]
D 10-23 00:20:46 instance.py:66] handler_to_instances: defaultdict(<class 'list'>, {})
D 10-23 00:20:46 instance.py:55] handlers: [<class 'sky.provision.gcp.instance_utils.GCPComputeInstance'>]
D 10-23 00:20:47 instance.py:66] handler_to_instances: defaultdict(<class 'list'>, {})
D 10-23 00:20:47 provisioner.py:333] Provision record:
D 10-23 00:20:47 provisioner.py:333] {
D 10-23 00:20:47 provisioner.py:333]   "provider_name": "gcp",
D 10-23 00:20:47 provisioner.py:333]   "region": "us-central1",
D 10-23 00:20:47 provisioner.py:333]   "zone": "us-central1-a",
D 10-23 00:20:47 provisioner.py:333]   "cluster_name": "test-new-gcp-2-9ce1",
D 10-23 00:20:47 provisioner.py:333]   "head_instance_id": "test-new-gcp-2-9ce1-head-97577999-compute",
D 10-23 00:20:47 provisioner.py:333]   "resumed_instance_ids": [],
D 10-23 00:20:47 provisioner.py:333]   "created_instance_ids": [
D 10-23 00:20:47 provisioner.py:333]     "test-new-gcp-2-9ce1-head-97577999-compute"
D 10-23 00:20:47 provisioner.py:333]   ]
D 10-23 00:20:47 provisioner.py:333] }
D 10-23 00:20:47 provisioner.py:333] Cluster info:
D 10-23 00:20:47 provisioner.py:333] {
D 10-23 00:20:47 provisioner.py:333]   "instances": {},
D 10-23 00:20:47 provisioner.py:333]   "head_instance_id": null,
D 10-23 00:20:47 provisioner.py:333]   "docker_user": null
D 10-23 00:20:47 provisioner.py:333] }
E 10-23 00:20:47 provisioner.py:494] *** Failed setting up cluster. ***
D 10-23 00:20:47 provisioner.py:495] Stacktrace:
D 10-23 00:20:47 provisioner.py:495] Traceback (most recent call last):
D 10-23 00:20:47 provisioner.py:495]   File "/Users/zhwu/Library/CloudStorage/OneDrive-Personal/AResource/PhD/Research/sky-computing/code/sky-experiment-dev/sky/provision/provisioner.py", line 486, in post_provision_runtime_setup
D 10-23 00:20:47 provisioner.py:495]     return _post_provision_setup(cloud_name,
D 10-23 00:20:47 provisioner.py:495]   File "/Users/zhwu/Library/CloudStorage/OneDrive-Personal/AResource/PhD/Research/sky-computing/code/sky-experiment-dev/sky/provision/provisioner.py", line 341, in _post_provision_setup
D 10-23 00:20:47 provisioner.py:495]     raise RuntimeError(f'Provision failed for cluster {cluster_name!r}. '
D 10-23 00:20:47 provisioner.py:495] RuntimeError: Provision failed for cluster 'test-new-gcp-2'. Could not find any head instance.
D 10-23 00:20:47 provisioner.py:495] 

```

Reply:
thanks! this should have been fixed

Reply:
actually we use `wait_instances` immediately after `run_instances` in AWS, so the whole process is not async. I am considering removing `wait_instances`

Reply:
conclusion: we should consider merging `wait_instances` into `run_instances`